lsm-tree.cabal

```diff
@@ -52,45 +52,103 @@ description:
-  As such, the worst-case complexity of basic operations refer to the number of disk I\/O actions.
-
-  TODO: Describe the time complexity of the basic operations.
+  The worst-case behaviour of the library is described using [big-O notation](http://en.wikipedia.org/wiki/Big_O_notation).
+  The documentation provides two measures of complexity:
+
+  *   The time complexity of operations is described in terms of the number of disk I\/O operations and referred to as the disk I\/O complexity.
+      In practice, the time of the operations on LSM-trees is dominated by the number of disk I\/O actions.
+  *   The space complexity of tables is described in terms of the in-memory size of an LSM-tree table.
+      Both the in-memory and on-disk size of an LSM-tree table scales linearly.
+      However, the in-memory size of an LSM-tree table is smaller than its on-disk size by a significant constant.
+      This is discussed in detail below, under [In-memory size of tables](#performance_size).
+
+  The complexities are described in terms of the following variables and constants:
+
+  *   The variable \(n\) refers to the number of /physical/ table entries.
+      A /physical/ table entry is any key–operation pair, e.g., @Insert k v@ or @Delete k@, whereas a /logical/ table entry is determined by all physical entries with the same key.
+      If the variable \(n\) is used to describe the complexity of an operation that involves multiple tables, it refers to the sum of all table entries.
+  *   The constant \(T\) refers to the size ratio of the table, which is a configuration parameter.
+
+  === Disk I\/O cost of operations #performance_time#
+
+  The following table summarises the cost of the operations on LSM-trees
+  measured in the number of disk I\/O operations.
+  If the cost depends on the merge policy, the table contains one entry for each merge policy.
+  Otherwise, the merge policy is listed as N\/A.
+
+  +----------+------------------------+----------------------------+------------------------------+
+  | Resource | Operation              | Merge policy               | Cost in disk I\/O operations |
+  +==========+========================+============================+==============================+
+  | Session  | Create\/Open           | N\/A                       | \(O(1)\)                     |
+  +----------+------------------------+----------------------------+------------------------------+
+  |          | Close                  | N\/A                       | \(O(t \: \log_T \: n)\)*     |
+  +----------+------------------------+----------------------------+------------------------------+
+  | Table    | Create                 | N\/A                       | \(O(1)\)                     |
+  +----------+------------------------+----------------------------+------------------------------+
+  |          | Close                  | N\/A                       | \(O(\log_T \: n)\)           |
```

Comment:
Functions like this where the complexity depends on the number of runs in a table, should also depends on the merge policy: if $\log_T n$ approximates the number of levels, then with tiering there are at most $T-1$ runs per level, and with levelling at most $1$ per level

Reply:
Regardless, the number of runs is proportional to the number of levels, so the big-O complexity is the same, right?

Reply:
I'm a little bit hazy on the specific rules for big-O complexity, but I tend to consider $T$ to be a parameter to the algorithm (even if it is currently fixed to 4), in which case it shouldn't be considered a constant value I think

If $T$ is considered constant, shouldn't $\log_T$ then equivalent under big-O to $\log_2$?